### ISSUE TITLE
Ability to await for execution of all @Async methods

### ISSUE THREAD

**Issue by musketyr** [micronaut-projects/micronaut-core#7858]
### Feature description

We are trying to move non-business related code into async event listeners but when we are running AWS functions then some of the methods annotated with `@Async` are not executed because the Lambda is put into sleep mode before the code is finished.  We we able to make this work with a following snippet:

```groovy
@Named(TaskExecutors.SCHEDULED) @Inject ExecutorService executorService

executorService.shutdown()
executorService.awaitTermination(5, TimeUnit.SECONDS)
```

This is pretty primitive approach and I wonder if there could be some bean which would provide such an interface which is more reliable (e.g. will work out of box with any other custom `ExecutorService`). I'm also bit worried what will happen if the Lambda is reused.

**Comment by dstepanov** [comment]
Shutting down the executor service will break the runtime when the lambda is reused.
You might want to create a custom `ExecutorService` that will collect all the tasks and await their completion before the lambda is done.

**Comment by musketyr** [comment]
Thanks for fast response. Is this something which can be done directly in Micronaut libraries, ideally somewhere inside `AsyncInterceptor`? I believe this would be helpful for every Micronaut user.

**Comment by dstepanov** [comment]
Maybe somewhere in the Lambda project, pretty much you just need a custom scheduler:
```java
@Singleton
@Named("LambdaScheduler")
class LambdaSchedulerScheduler implements ExecutorService {
   ///...

   void awaitTasksCompletion() { ... }

}

@Async("LambdaScheduler")
void myAsyncMethod() {...}

void someLambdaWork() {
  //...
  lambdaSchedulerScheduler.awaitTasksCompletion();
}

```
I don't know much about Lamdas implementation, but assuming the context is reused. I haven't found any events that would be automatically triggered at the end of the invocation, those would have been helpful to do the await.

**Comment by musketyr** [comment]
Yes, the context is reused and you were right about the `shutdown`. I have created a test function and it looks like your suggestion with registering the futures works. So the question is if you would like to add something like this directly inside `micronaut-function` or should I create our own standalone library.

https://github.com/musketyr/micronaut-function-async-support/tree/master/src/main/java/com/example/async

**Comment by dstepanov** [comment]
It might be a good idea to have it first as a part of `micronaut-aws` and then maybe to move it to `micronaut-function`.  WDYT @graemerocher 
Also, it would be nice to introduce a new event being triggered after the lambda function invocation, and this new scheduler can have a listener and do the awaiting automatically.

BTW: you can use `CompletableFuture<Void> runAsync(Runnable runnable, Executor executor)` in the scheduler and `CompletableFuture<Void> allOf` to implement await.

**Comment by musketyr** [comment]
> Also, it would be nice to introduce a new event being triggered after the lambda function invocation, and this new scheduler can have a listener and do the awaiting automatically.
>
If we stick directly to the `io.micronaut.function.aws.MicronautRequestHandler` and not more generic `io.micronaut.function.executor.FunctionInitializer` (which we are historically using - I believe one time it was a favorite way how to write handlers) then yes, it would be good if a sync event was fired at the end of execution of `handleRequest` method.

```java
    @Override
    public final O handleRequest(I input, Context context) {
        HandlerUtils.configureWithContext(this, context);
        if (!inputType.isInstance(input)) {
            input = convertInput(input);
        }
        O output = this.execute(input);
        publishExecutionFinishedEvent(input, output);
        return output;
    }
```

The same applies for `io.micronaut.function.aws.MicronautRequestStreamHandler`. It should not be a big deal for us to migrate our functions. Once there is a simple way for post-execute hooks then we can simplify a lot of our code, including for example Sentry Integration where we need to wait for flushing the events.


 
> BTW: you can use `CompletableFuture<Void> runAsync(Runnable runnable, Executor executor)` in the scheduler and `CompletableFuture<Void> allOf` to implement await.

Thanks for a tip!

**Comment by musketyr** [comment]
I have started the work on the event here https://github.com/micronaut-projects/micronaut-aws/pull/1433

**Comment by musketyr** [comment]
Just a few more details why we would like to keep `@Async` even in Lambda environment. We simply have a services which are shared between Micronaut applications and Micronaut functions and we don't want to create pairs of duplicates for sync/async invocations.

**Comment by dstepanov** [comment]
In that case, you might want to wrap all the executors in the Lamba environment using a bean listener like this https://github.com/micronaut-projects/micronaut-core/blob/3.6.x/context/src/main/java/io/micronaut/scheduling/instrument/ExecutorServiceInstrumenter.java and await the completion.

**Comment by musketyr** [comment]
Thanks for the link! It may be really handly!

**Comment by musketyr** [comment]
Just one radical idea. Would it be possible to ignore `@Async` annotation/interceptor and make `publishEventAsync` call synchronous when AWS Lambda function environment is detected?

**Comment by prashantcodecraft** [comment]
👍 

Any way this work can be progressed. 

We are facing same issue with Application hosted as Lambda Functions where is the service response stops lambda execution and do not wait for ASYNC methods to finish.

**Comment by ojebari-ma** [comment]
Can you please confirm that this is still an issue in the latest version of Micronaut?

**Comment by musketyr** [comment]
the problem is still there. I wasn't able to find any solution in the framework that would address issues with JVM hibernation. internally, we execute all code synchronously when we execute code in the function environment as a workaround.

```
@Singleton
@Replaces(AsyncInterceptor.class)
@Requires(condition = NoAsyncCondition.class)
public class NoAsyncInterceptor implements MethodInterceptor<Object, Object> {

    public static final String NO_ASYNC = "no-async";
    public static final String FORCE_ASYNC = "force-async";

    @Override
    public Object intercept(MethodInvocationContext<Object, Object> context) {
        return context.proceed();
    }

}

public class NoAsyncCondition implements Condition {

    private static final List<String> ENABLED_ENVS = List.of(NoAsyncInterceptor.NO_ASYNC, Environment.TEST, Environment.FUNCTION, "lambda");

    @Override
    public boolean matches(ConditionContext context) {
        BeanContext beanContext = context.getBeanContext();

        if (beanContext instanceof ApplicationContext) {
            ApplicationContext applicationContext = (ApplicationContext) beanContext;
            Set<String> envs = applicationContext.getEnvironment().getActiveNames();

            if (envs.contains(NoAsyncInterceptor.FORCE_ASYNC)) {
                return false;
            }

            return ENABLED_ENVS.stream().anyMatch(envs::contains);
        }

        return false;
    }

}
```

**Comment by graemerocher** [comment]
Triage: this is a valid async-lifecycle enhancement, and the reporter confirms it still affects Lambda-style runtimes. @Async currently submits work through an executor without a framework-level completion registry, so the host can finish before fire-and-forget work completes. A practical design would expose an await/drain hook for framework-managed @Async tasks, document its limits for arbitrary custom executors, and integrate it with function invocation shutdown; the synchronous interceptor replacement remains a workaround.
